static_frame/core/index_hierarchy.py

```diff
@@ -1794,70 +2364,173 @@ class IndexHierarchyGO(IndexHierarchy):
+            )
+
+        label_indexers = []
+
+        for depth, label_at_depth in enumerate(value):
+            if label_at_depth in self._indices[depth]:
+                label_index = self._indices[depth]._loc_to_iloc(label_at_depth)
+            else:
+                label_index = len(self._indices[depth])
+                self._indices[depth].append(label_at_depth)
 
-    def extend(self, other: IndexHierarchy) -> None:
+            label_indexers.append(label_index)
+
+        for i, label_index in enumerate(label_indexers):
+            self._indexers[i] = np.append(self._indexers[i], label_index)
+            self._indexers[i].flags.writeable = False
+
+        # No need to ensure uniqueness! It's already been checked.
+        self._blocks = self._create_blocks_from_self()
+        self._values = self._blocks.values
+
+    def extend(self: IHGO,
+            other: IndexHierarchy,
+            ) -> None:
         '''
-        Extend this IndexHiearchy in-place
+        Extend this IndexHierarchy in-place
         '''
-        self._levels.extend(other._levels)
-        self._recache = True
+        mask_flags = np.array([1, -1], dtype=DTYPE_INT_DEFAULT) # False = n, True = -n
 
-    def __copy__(self: IH) -> IH:
+        # TODO: How to handle corrupted instance state if an exception is raised?
+        for depth, (self_index, other_index) in enumerate(
+                zip(self._indices, other._indices)
+                ):
+            intersection = self_index.intersection(other_index)
+            if not intersection.size:
+                # Easy case; all of the values at this level are not in self, so we
+                # can easily extend both the index & the indexer
+                del intersection
+
+                starting_len = len(self_index)
+
+                self_index.extend(other_index)
+
+                new_indexer = np.hstack(
+                    (self._indexers[depth], other._indexers[depth] + starting_len)
+                    )
+                new_indexer.flags.writeable = False
+                self._indexers[depth] = new_indexer
+                continue
+
+            if len(intersection) == len(self_index) == len(other_index):
+                # Easy case; every single label in the incoming index is also in self, so we
+                # just have to extend the indexer
+                del intersection
+
+                if self_index.equals(other_index):
+                    # Easiest case, no need to remap the indexer since the order is the same
+
+                    new_indexer = np.hstack(
+                        (self._indexers[depth], other._indexers[depth])
+                    )
+                    new_indexer.flags.writeable = False
+                    self._indexers[depth] = new_indexer
+                    continue
+
+                # Same labels, but different order. We have to remap the indexers.
+                indexer_remap = other_index.iter_label().apply(self_index._loc_to_iloc)
```

Comment:
As in other contexts, there is likely a lower-level interface to doing the `iter_label().apply()` here.